api/v1alpha1/pkisecretengineconfig_types.go

```diff
@@ -0,0 +1,535 @@
+	return string(p.Spec.Path) + "/" + p.Spec.Type + "/generate/" + p.Spec.PrivateKeyType
+}
+
+func (p *PKISecretEngineConfig) GetConfigUrlsPath() string {
+	return string(p.Spec.Path) + "/config/urls"
+}
+
+func (p *PKISecretEngineConfig) GetConfigCrlPath() string {
+	return string(p.Spec.Path) + "/config/crl"
+}
+
+func (p *PKISecretEngineConfig) GetSignIntermediatePath() string {
+	return string(p.Spec.InternalSign.Name) + "/root/sign-intermediate"
+}
+
+func (p *PKISecretEngineConfig) GetIntermediateSetSignedPath() string {
+	return string(p.Spec.Path) + "/intermediate/set-signed"
+}
+
+func (p *PKISecretEngineConfig) GetGeneratedStatus() bool {
+	return p.Status.Generated
+}
+
+func (p *PKISecretEngineConfig) SetGeneratedStatus(status bool) {
+	p.Status.Generated = status
+}
+
+func (p *PKISecretEngineConfig) CreateExported(context context.Context, secret *vault.Secret) (bool, error) {
+	log := log.FromContext(context)
+	payload := p.GetExportedPayload(secret.Data)
+	exported := p.Spec.PrivateKeyType == "exported"
+
+	if exported {
+		kubeClient := context.Value("kubeClient").(client.Client)
+		kubeSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      p.Name,
+				Namespace: p.Namespace,
+				Labels: map[string]string{
+					"redhatcop.redhat.io/pkisecretengineconfigs": p.Name,
+				},
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(p, p.GroupVersionKind()),
+				},
+			},
+			StringData: payload,
+		}
+
+		err := kubeClient.Create(context, kubeSecret)
+		if err != nil {
+			log.Error(err, "unable to create exported secret", "path", kubeSecret)
+			return false, err
+		}
+	}
+
+	if p.Spec.Type == "intermediate" {
+		p.Spec.PKIIntermediate.CSR = payload["csr"]
+	}
+
+	return exported, nil
+}
+
+func (p *PKISecretEngineConfig) SetExportedStatus(status bool) {
+	p.Status.Exported = status
+}
+
+func (p *PKISecretEngineConfig) SetIntermediate(context context.Context) error {
+	if p.Spec.Type == "intermediate" {
+		log := log.FromContext(context)
+		vaultClient := context.Value("vaultClient").(*vault.Client)
+
+		if p.Spec.InternalSign != nil && p.Spec.InternalSign.Name != "" {
+			secret, err := vaultClient.Logical().Write(p.GetSignIntermediatePath(), p.GetSignIntermediatePayload())
+			if err != nil {
+				log.Error(err, "unable to write object at", "path", p.GetIntermediateSetSignedPayload())
+				return err
+			}
+			p.setSignedIntermediate(secret.Data["certificate"].(string))
+		}
+
+		_, err := vaultClient.Logical().Write(p.GetIntermediateSetSignedPath(), p.GetIntermediateSetSignedPayload())
+		if err != nil {
+			log.Error(err, "unable to write object at", "path", p.GetIntermediateSetSignedPayload())
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *PKISecretEngineConfig) GetExportedPayload(data map[string]interface{}) map[string]string {
+
+	payload := map[string]string{}
+
+	if p.Spec.Type == "root" {
+		payload["issuing_ca"] = toString(data["issuing_ca"])
```

Comment:
ok I see where you are using to string. why did you have to do it this way?

Reply:
what is the better approach? can't sign an interface to a string.